File: backend/src/Application/Emails/ReplyDrafts/MailReplyDrafting.cs
// Copyright © 2026 Krzysztof Kasprowicz
// Licensed under the GNU Affero General Public License, Version 3. See LICENSE in the project root for license information.
// Project repository: https://github.com/Krzysztof318/MailFathom

using MailFathom.Application.Access;
using MailFathom.Application.Emails.Mailboxes;
using MailFathom.Application.Retrieval.AskMail;
using MailFathom.Application.SensitiveContent.Detection;
using MailFathom.Application.SensitiveContent.Egress;
using MailFathom.Domain.Access;
using MailFathom.Domain.Emails;

namespace MailFathom.Application.Emails.ReplyDrafts;

/// <summary>Drafts a reply to one of the acting user's messages, out of the conversation and the way they write.</summary>
/// <remarks>
/// <para>
/// <b>It produces a local artifact and takes no act.</b> Nothing here sends, queues, or writes to a mail server, and
/// nothing is stored: the reply comes back as text the person owns, and saving it as a draft or sending it stays where
/// those acts already are, behind somebody confirming them.
/// </para>
/// <para>
/// The correspondence is read under the same scope the conversation screen reads it under — every account this user
/// holds, no folder narrowing, junk included — because a reply is answered across the exchange rather than inside a
/// folder. A message the scope does not admit is a message this use case has none of, so the drafting answers exactly
/// as it does for a message nobody holds.
/// </para>
/// <para>
/// The style is the account's own sent mail and nothing else, and the read that fetches it is the read that resolved
/// the answered message, which is what keeps one person's manner out of another account's reply. An operator who
/// turned that derivation off leaves the bound at zero: no sent mail is read, and the draft is written from the
/// conversation alone.
/// </para>
/// <para>
/// Both halves of the privacy posture are here rather than in the writer. What goes out to a provider is guarded by
/// the writer at the prompt; what comes back is guarded here before it reaches a client, because a draft is written out
/// of somebody's correspondence and carries whatever that correspondence carried.
/// </para>
/// </remarks>
public sealed class MailReplyDrafting
{
    /// <summary>How many of the conversation's most recent messages one drafting is grounded in.</summary>
    /// <remarks>
    /// A reply answers where a conversation got to, so the recent end of it is what a draft is written from; the
    /// opening of a long exchange is what its derived state is for. The bound is a constant rather than a setting
    /// because it decides how long one drafting takes rather than what a deployment is — what an operator declares
    /// about the spend is the answering period's ceilings, which every drafting is admitted against.
    /// </remarks>
    public const int MaximumMessages = 20;

    /// <summary>How much of one of those messages travels with the drafting.</summary>
    /// <remarks>The bound a conversation's own derivation reads a message under, because what is being bounded is the same text: what a message added, with the history it quoted already trimmed off.</remarks>
    public const int MaximumCharactersPerMessage = 4_000;

    /// <summary>How many of the person's own recent sent messages a style is derived from.</summary>
    /// <remarks>
    /// Enough for a manner to be visible — how they open, how they sign off, whether they write in paragraphs or in
    /// lines — and few enough that drafting one reply never turns into reading somebody's outbox. Every one of them is
    /// mail this deployment already holds and never leaves it except as this drafting's own prompt.
    /// </remarks>
    public const int MaximumStyleMessages = 6;

    /// <summary>How much of one sent message travels for that derivation.</summary>
    /// <remarks>Shorter than a message of the conversation, because what is being read out of it is a manner rather than a meaning, and the opening of a reply is where a manner shows.</remarks>
    public const int MaximumStyleCharactersPerMessage = 1_200;

    private readonly IReplyDraftSourceReader sourceReader;
    private readonly IReplyDraftWriter writer;
    private readonly MailboxScopeResolver scopeResolver;
    private readonly SensitiveContentEgressGuard egressGuard;
    private readonly AccessAuthorization authorization;
    private readonly bool derivesStyleFromSentMail;

    /// <summary>Initializes the drafting.</summary>
    /// <param name="sourceReader">Reads the conversation being answered and the manner its account writes in.</param>
    /// <param name="writer">Writes the reply, which is the one part of this that reaches a provider.</param>
    /// <param name="scopeResolver">Decides which accounts and folders the conversation is read across.</param>
    /// <param name="egressGuard">Scans what the draft publishes to a client, where this deployment scans anything.</param>
    /// <param name="authorization">Enforces the permission this drafting is behind.</param>
    /// <param name="derivesStyleFromSentMail">Whether the deployment derives a manner from the account's own sent mail at all.</param>
    /// <exception cref="ArgumentNullException">Thrown when any dependency is <see langword="null" />.</exception>
    public MailReplyDrafting(
        IReplyDraftSourceReader sourceReader,
        IReplyDraftWriter writer,
        MailboxScopeResolver scopeResolver,
        SensitiveContentEgressGuard egressGuard,
        AccessAuthorization authorization,
        bool derivesStyleFromSentMail)
    {
        ArgumentNullException.ThrowIfNull(sourceReader);
        ArgumentNullException.ThrowIfNull(writer);
        ArgumentNullException.ThrowIfNull(scopeResolver);
        ArgumentNullException.ThrowIfNull(egressGuard);
        ArgumentNullException.ThrowIfNull(authorization);

        this.sourceReader = sourceReader;
        this.writer = writer;
        this.scopeResolver = scopeResolver;
        this.egressGuard = egressGuard;
        this.authorization = authorization;
        this.derivesStyleFromSentMail = derivesStyleFromSentMail;
    }

    /// <summary>Drafts a reply to one of the acting user's messages.</summary>
    /// <param name="request">Which message is being answered, and what its author asked the reply to say.</param>
    /// <param name="cancellationToken">Propagates caller cancellation.</param>
    /// <returns>
    /// The draft, <see cref="ReplyDraft.Nothing" /> where none could be written, or <see langword="null" /> where this
    /// user holds no such message.
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown when <paramref name="request" /> is <see langword="null" />.</exception>
    /// <exception cref="MailAnsweringBudgetExhaustedException">Thrown when this deployment has spent what it allows a provider for the period.</exception>
    /// <exception cref="SensitiveContentScannerUnavailableException">Thrown when a switched-on scanner could not establish what the draft carries, which withholds it rather than publishing it unscanned.</exception>
    /// <exception cref="PrincipalNotAuthorizedException">Thrown when the use case was reached by anything but a caller granted <see cref="MailFathomPermission.MailAsk" />.</exception>
    public async Task<ReplyDraft?> DraftAsync(ReplyDraftRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        // The grant that puts mail in front of a chat provider rather than the one that reads it, because that is what
        // this does: the conversation and a sample of the account's own sent mail leave this deployment, and the call
        // is charged to the same allowance a question is.
        this.authorization.RequirePermission(MailFathomPermission.MailAsk);

        using var actingFor = this.egressGuard.ActingFor(this.scopeResolver.User);

        var scope = this.scopeResolver.ReadableScope([], [], JunkMailInclusion.Included);

        if (scope.AccountIds.Count is 0)
        {
            return null;
        }

        var sources = await this.sourceReader.ReadSourcesAsync(
            request.AnsweredEmailId,
            scope,
            this.Bounds(),
            cancellationToken);

        if (sources is null)
        {
            return null;
        }

        if (sources.Messages.Count is 0)
        {
            // A conversation this deployment has stored no readable text for. There is nothing to ground a reply in,
            // and asking a provider to write one anyway would produce a fluent message resting on nothing at all.
            return ReplyDraft.Nothing;
        }

        var brief = new ReplyDraftBrief(
            sources,
            Bounded(request.Selection, ReplyDraftRequest.MaximumSelectionLength),
            Bounded(request.Instruction, ReplyDraftRequest.MaximumInstructionLength));

        var draft = await this.writer.WriteAsync(brief, cancellationToken);

        return draft.WasWritten ? await this.GuardedAsync(draft, cancellationToken) : draft;
    }

    /// <summary>Cuts what somebody typed down to what one drafting may carry of it.</summary>
    /// <remarks>
    /// Shortened rather than refused, because the boundary above already refuses in the words of the bound and this is
    /// what makes the rule hold for every caller rather than for the one that was written against it.
    /// </remarks>
    private static string? Bounded(string? typed, int maximumLength) =>
        string.IsNullOrWhiteSpace(typed)
            ? null
            : MailTextBounds.TruncateAtTextElementBoundary(typed.Trim(), maximumLength);

    private ReplyDraftBounds Bounds() => new(
        MaximumMessages,
        MaximumCharactersPerMessage,
        this.derivesStyleFromSentMail ? MaximumStyleMessages : 0,
        MaximumStyleCharactersPerMessage);

    /// <summary>Scans everything the draft would publish, under the point this surface is read on.</summary>
    /// <remarks>
    /// One report for the draft rather than one per claim, because the draft is what a composer waits for. The body and
    /// the claims are offered; the addresses are not, being values this deployment resolved out of its own store rather
    /// than text a producer wrote, and a redacted address would be one nobody could send to.
    /// </remarks>
    private async Task<ReplyDraft> GuardedAsync(ReplyDraft draft, CancellationToken cancellationToken)
    {
        if (!this.egressGuard.IsActive)
        {
            return draft;
        }

        using var scan = this.egressGuard.BeginGuardedOperation(
            SensitiveContentEgressPoint.ClientReplyDraft,
            cancellationToken);

        var body = await this.egressGuard.GuardAsync(
            SensitiveContentEgressPoint.ClientReplyDraft,
            draft.Body,
            cancellationToken);

        var claims = new List<ReplyDraftClaim>(draft.Claims.Count);

        foreach (var claim in draft.Claims)
        {
            claims.Add(ReplyDraftClaim.Create(
                await this.egressGuard.GuardAsync(
                    SensitiveContentEgressPoint.ClientReplyDraft,
                    claim.Text,
                    cancellationToken),
                claim.Sources));
        }

        scan.Completed();

        return ReplyDraft.Written(body, claims, draft.ProposedRecipients);
    }
}

File: frontend/src/Client.Backend/src/mailReplyDrafting.ts
// Copyright © 2026 Krzysztof Kasprowicz
// Licensed under the GNU Affero General Public License, Version 3. See LICENSE in the project root for license information.
// Project repository: https://github.com/Krzysztof318/MailFathom

import { failed, failureReasonForStatus, read, type ClientResult } from './failure';
import { asRecord } from './json';
import { headersFor, routeFor, type ClientSession } from './session';
import { spanned } from './telemetry';
import { send, type MailFathomTransport } from './transport';

// The draft somebody is about to write, composed by the deployment out of the conversation it answers or out of what
// they typed. What comes back is a proposal rather than a message: nothing is sent, nothing is filed, and the text
// arrives in a composer its author edits, discards, or saves through the routes that already do those things.
//
// **Two routes on one address, because a composer has two questions and only one of them costs anything.** The read
// says whether this deployment drafts at all, which is what decides whether the affordance is drawn before anybody
// presses it — a control promising a draft over a deployment that writes none fails a person at the one moment they
// trusted it. The write drafts one and costs a provider call.
//
// **Three answers rather than a value and a failure**, because a composer does something different with each. A draft
// is put on the screen. A deployment that drafts none, and one whose provider could not be reached, are the composer
// as it was and are not a failure anybody can act on — which is why the service answers both `200` and this module
// reports them as one outcome rather than inventing a difference. An allowance the deployment has already spent is the
// one that travels, because somebody pressed a button the operator has paid the last of the budget for and being told
// nothing is what would send them pressing it again.
//
// Everything here is derived from somebody's own mail and is personal data of the same standing, so none of it is
// logged, kept, or put in a span attribute; the span names the route and carries nothing of the draft.

/** The route a reply is drafted at, relative to the client prefix. */
export const mailReplyDraftingRoute = '/replies/drafting';

/** One message a claim rests on, spelled as every citation on this surface is spelled. */
export interface MailReplyDraftSource {
    readonly kind: 'email';
    readonly email: string;
}

/** One thing the drafted message asserts, and what the correspondence does or does not say about it. */
export interface MailReplyDraftClaim {
    /** The assertion, as one sentence of the draft. */
    readonly text: string;

    /**
     * Whether the correspondence backs it.
     *
     * Read rather than derived from the sources being empty: what it means is a promise the deployment makes about the
     * draft, and a client deciding it from an array's length would be re-deciding, per client, what an unsupported
     * claim is.
     */
    readonly supported: boolean;

    /** The messages backing it, best first, and empty for a claim the correspondence does not carry. */
    readonly sources: readonly MailReplyDraftSource[];
}

/** One person the draft proposes to reach, every one of whom the conversation itself names. */
export interface MailReplyDraftRecipient {
    readonly address: string;
    readonly displayName: string | null;
}

/** A draft the deployment wrote, as a composer puts it on the screen. */
export interface MailReplyDraft {
    readonly outcome: 'drafted';

    /** The draft as plain text, which is what the composer's body is set from. */
    readonly body: string;

    /** What the draft asserts, in the order it was written. */
    readonly claims: readonly MailReplyDraftClaim[];

    /** Who the draft proposes to reach, shown as proposals and added to nothing on anybody's behalf. */
    readonly proposedRecipients: readonly MailReplyDraftRecipient[];
}

/** What a drafting answered with, which is a draft or one of the two reasons there is none. */
export type MailReplyDrafting =
    MailReplyDraft | { readonly outcome: 'notDrafted' } | { readonly outcome: 'allowanceSpent' };

/** The answer a deployment that writes no draft gives, which leaves the composer as it was. */
export const nothingDrafted: MailReplyDrafting = { outcome: 'notDrafted' };

/** What one drafting asks for: the message being answered, if any, and what its author wants said. */
export interface MailReplyDraftRequest {
    /** The message being answered, or `null` where the message being written answers none. */
    readonly answeredEmailId: string | null;

    /** The part of the correspondence being answered, or `null` to answer it as a whole. */
    readonly selection: string | null;

    /** What the draft should say, or `null` where the person asked for nothing in particular. */
    readonly instruction: string | null;
}

/**
 * The greatest instruction the service reads.
 *
 * The client's own copy of the service's bound, so a composer shortens what it sends rather than putting a request
 * together that the deployment will refuse with a sentence nobody asked for.
 */
export const longestDraftInstruction = 1_000;

/** The greatest selection the service reads, on the same terms as the instruction above. */
export const longestDraftSelection = 4_000;

// A draft is a message and a handful of short claims, so this is generous against what the route answers and exists
// for the answer that is not a draft at all.
const longestDraftAnswer = 128 * 1024;

// What one answer may carry before the draft is refused unread. The service bounds each of these itself; these are the
// client's own copy of the bound, checked while the collections are walked rather than after.
const longestIdentity = 256;
const longestBody = 32_000;
const longestClaimText = 4_096;
const longestAddress = 512;
const mostClaims = 32;
const mostSourcesPerClaim = 16;
const mostProposedRecipients = 32;

/**
 * Reads whether this deployment drafts a message at all, which is what decides whether the affordance is offered.
 *
 * @param session The address to reach and the finished header value to present.
 * @param transport How the request goes out.
 * @returns Whether a draft is written, or why the answer never arrived.
 */
export function draftsMailReplies(
    session: ClientSession,
    transport: MailFathomTransport,
): Promise<ClientResult<boolean>> {
    return spanned(`GET ${mailReplyDraftingRoute}`, async () => {
        const response = await send(transport, {
            method: 'GET',
            path: routeFor(session, mailReplyDraftingRoute),
            headers: headersFor(session),
            longestAnswer: longestDraftAnswer,
        });

        if (response === null) {
            return failed('unavailable', null);
        }

        if (response.status !== 200) {
            return failed(failureReasonForStatus(response.status), response.status);
        }

        const record = parsed(response.body);
        const drafts = record === null ? null : record['draftsReplies'];

        return typeof drafts === 'boolean' ? read(drafts) : failed('unreadable', response.status);
    });
}

/**
 * Drafts one message out of the conversation it answers, or out of what its author asked for.
 *
 * @param session The address to reach and the finished header value to present.
 * @param transport How the request goes out.
 * @param request The message being answered, if any, and the two texts beside it.
 * @returns The draft, the two shapes of having none, or why the answer never arrived.
 */
export function draftMailReply(
    session: ClientSession,
    transport: MailFathomTransport,
    request: MailReplyDraftRequest,
): Promise<ClientResult<MailReplyDrafting>> {
    return spanned(`POST ${mailReplyDraftingRoute}`, async () => {
        const response = await send(transport, {
            method: 'POST',
            path: routeFor(session, mailReplyDraftingRoute),
            headers: { ...headersFor(session), 'Content-Type': 'application/json' },
            body: JSON.stringify({
                answeredEmailId: request.answeredEmailId,
                selection: shortened(request.selection, longestDraftSelection),
                instruction: shortened(request.instruction, longestDraftInstruction),
            }),
            longestAnswer: longestDraftAnswer,
        });

        if (response === null) {
            return failed('unavailable', null);
        }

        // The one non-success status this route answers that a composer draws rather than reports as something going
        // wrong, because the operator has spent the allowance and the person pressing the control has to be told so.
        if (response.status === 429) {
            return read({ outcome: 'allowanceSpent' });
        }

        // The message the composer was opened over is gone. A draft of it is not something to retry and not something
        // to sign in again for, so it is the composer as it was, which is the answer a deployment drafting none gives.
        if (response.status === 404) {
            return read(nothingDrafted);
        }

        if (response.status !== 200) {
            return failed(failureReasonForStatus(response.status), response.status);
        }

        const draft = parseDrafting(response.body);

        return draft === null ? failed('unreadable', response.status) : read(draft);
    });
}

/** Cuts a typed text to what the service reads, so a composer sends a request the deployment will answer. */
function shortened(typed: string | null, longest: number): string | null {
    const written = typed?.trim() ?? '';

    return written.length === 0 ? null : written.slice(0, longest);
}

function parsed(body: string): Record<string, unknown> | null {
    try {
        return asRecord(JSON.parse(body));
    } catch {
        return null;
    }
}

function parseDrafting(body: string): MailReplyDrafting | null {
    const record = parsed(body);
    if (record === null) {
        return null;
    }

    if (record['drafted'] !== true) {
        return record['drafted'] === false ? nothingDrafted : null;
    }

    const draftBody = record['body'];
    if (typeof draftBody !== 'string' || draftBody.length === 0 || draftBody.length > longestBody) {
        return null;
    }

    const claims = parseClaims(record['claims']);
    const proposedRecipients = parseRecipients(record['proposedRecipients']);

    return claims === null || proposedRecipients === null
        ? null
        : { outcome: 'drafted', body: draftBody, claims, proposedRecipients };
}

function parseClaims(value: unknown): readonly MailReplyDraftClaim[] | null {
    if (!Array.isArray(value) || value.length > mostClaims) {
        return null;
    }

    const claims: MailReplyDraftClaim[] = [];
    for (const written of value) {
        const record = asRecord(written);
        const text = record?.['text'];
        const supported = record?.['supported'];

        if (typeof text !== 'string' || text.length === 0 || text.length > longestClaimText) {
            return null;
        }

        if (typeof supported !== 'boolean') {
            return null;
        }

        const sources = parseSources(record?.['sources']);
        if (sources === null) {
            return null;
        }

        claims.push({ text, supported, sources });
    }

    return claims;
}

function parseSources(value: unknown): readonly MailReplyDraftSource[] | null {
    if (!Array.isArray(value) || value.length > mostSourcesPerClaim) {
        return null;
    }

    const sources: MailReplyDraftSource[] = [];
    for (const written of value) {
        const record = asRecord(written);
        const email = record?.['email'];

        if (record?.['kind'] !== 'email' || typeof email !== 'string' || email.length === 0) {
            return null;
        }

        if (email.length > longestIdentity) {
            return null;
        }

        sources.push({ kind: 'email', email });
    }

    return sources;
}

function parseRecipients(value: unknown): readonly MailReplyDraftRecipient[] | null {
    if (!Array.isArray(value) || value.length > mostProposedRecipients) {
        return null;
    }

    const recipients: MailReplyDraftRecipient[] = [];
    for (const written of value) {
        const record = asRecord(written);
        const address = record?.['address'];
        const displayName = record?.['displayName'] ?? null;

        if (typeof address !== 'string' || address.length === 0 || address.length > longestAddress) {
            return null;
        }

        if (displayName !== null && (typeof displayName !== 'string' || displayName.length > longestAddress)) {
            return null;
        }

        recipients.push({ address, displayName });
    }

    return recipients;
}
